### PR TITLE
Added character matrix layer functionality

### PR DESCRIPTION
**Pull request by mattjones315**
Added the ability to store an arbitrary number of character matrices in the `CassiopeiaTree` object using layers. This is motivated by the use of layers in `AnnData` objects, which are useful for storing expression matrices with different normalization routines.

Currently, users will have the ability to store different layers and use these layers for inferring ancestral states or reconstructing trees. A potential option would be to add a `layer` option to simulators, allowing new character matrices to be created on top of `CassiopeiaTrees` without overwriting existing data. This might be useful, for example, with simulators that add dropout to the object. Another use case would be if the user had different filtering options they'd like to try out, or different ways of resolving ambiguous characters.

### CODE REVIEW

File: cassiopeia/solver/HybridSolver.py
Comment (Lioscro) [inline cassiopeia/solver/HybridSolver.py:11]: Same here

File: cassiopeia/solver/ILPSolver.py
Comment (Lioscro) [inline cassiopeia/solver/ILPSolver.py:7]: and here

File: cassiopeia/data/CassiopeiaTree.py
Comment (Lioscro) [inline cassiopeia/data/CassiopeiaTree.py:130]: Do you think it would be useful to have some kind of function called `freeze_character_matrix`, which freezes the current dataframe in `cassiopeia_tree.character_matrix` into some layer?
(of course, this can be done manually by setting `cassiopeia_tree.layers['layer_name'] = cassiopeia_tree.character_matrix.copy()`)

**Review by Lioscro — CHANGES_REQUESTED** [review]
This looks awesome! I think having layers will greatly reduce the need to constantly make copies of the entire tree and make it much easier to keep track of different versions of the character matrix.

File: cassiopeia/solver/GreedySolver.py
Comment (mattjones315) [reply cassiopeia/solver/GreedySolver.py:6]: Doesn't look like it - thanks for pointing that out!

File: cassiopeia/data/__init__.py
Comment (mattjones315) [reply cassiopeia/data/__init__.py:4]: Yeah, this is a good point - I'll remove it from the `__init__`

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py:130]: I like this idea -- I'll add this functionality. Thanks for the suggestion!

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py:224]: This point, related to the one above, is fantastic. My intention is that a user can use _any_ character matrix of the tree for downstream analysis (which I think is clear from the motivation of creating the `Layers` class). And to a degree you're right, `character_matrix` does represent `current_character_matrix`, but I don't necessarily want to overwrite the existing `character_matrix` attribute every time you reset the leaves of the tree with new character states. In the same way, the desired behavior here was to not necessarily rewrite the states of the leaves when you set a new default character matrix.

I think this would be more clear though if we decouple setting the leaves of the tree from the character matrix setting. This would mean that a user would have to call `initialize_character_states_at_leaves` after setting the `character_matrix` attribute - something like:

```
tree.character_matrix = <character_matrix>
tree.initialize_character_states_at_leaves()
``` 

Or, if there is another layer available:

```
tree.set_character_states_at_leaves(layer='modified')
```

What do you think of this behavior?

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py]: See response above.

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py:129]: Implemented in latest commit.

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py]: See response below.

File: cassiopeia/data/CassiopeiaTree.py
Comment (Lioscro) [reply cassiopeia/data/CassiopeiaTree.py]: I actually really like your suggestion. I always think being explicit (and doing less things implicitly, like setting the character states at the leaves when setting the character matrix) is better. Why don't we go ahead and decouple the character matrix with the character states at the leaves?

File: cassiopeia/data/CassiopeiaTree.py
Comment (Lioscro) [reply cassiopeia/data/CassiopeiaTree.py]: I was referring to renaming this function to just `character_matrix` with a `@character_matrix.setter` decorator, which would call the function whenever someone assigns to `tree.character_matrix`, instead of having to call `tree.set_character_matrix()`.

Comment (Lioscro) [reply cassiopeia/data/CassiopeiaTree.py]: Whoops. Wasn't looking at the latest commit. haha

**Review by Lioscro — APPROVED** [review]
LGTM once we get the `set_character_states_at_leaves` flushed out.
One thing to note: you may want to rebase with the upstream (`refactor`) branch just before you merge and make sure that all the tests are passing, since this PR introduces a backwards-incompatible change in character matrix bookkeeping.

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py]: I agree - this all sounds great! I'm trying to decide if it would make the most sense to set the attribute `character_matrix` when you call `set_character_states_at_leaves(character_matrix=cm)`. This is why I removed this option initially, I can't decide what the best behavior would be.

**Comment by mattjones315** [comment]
Thanks for the feedback @Lioscro -- you mind taking one more look at the `set_character_states_at_leaves` functionality to make sure we're on the same page regarding expected behavior and functionality?

File: cassiopeia/data/CassiopeiaTree.py
Comment (Lioscro) [inline cassiopeia/data/CassiopeiaTree.py:262]: `sotred` -> `stored`

**Review by Lioscro — APPROVED** [review]
Looks good! one minor typo

File: cassiopeia/data/CassiopeiaTree.py
Comment (mattjones315) [reply cassiopeia/data/CassiopeiaTree.py:262]: Great, thanks!

**Pull request merged**
